flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/LocalBufferPool.java

```diff
@@ -231,18 +231,43 @@ private BufferBuilder toBufferBuilder(MemorySegment memorySegment) {
 
 	@Nullable
 	private MemorySegment requestMemorySegment(boolean isBlocking) throws InterruptedException, IOException {
+		// a flag to indicate whether the blocking request thread is
+		// waiting and needs to be notified
+		final AtomicBoolean waitingFlag = new AtomicBoolean(false);
 		MemorySegment segment = null;
 		synchronized (availableMemorySegments) {
 			returnExcessMemorySegments();
 
 			// fill availableMemorySegments with at least one element, wait if required
 			while (availableMemorySegments.isEmpty()) {
 				segment = requestMemorySegmentFromGlobal();
-				if (segment != null || !isBlocking) {
+				if (segment != null || !isBlocking || !availableMemorySegments.isEmpty()) {
```

Comment:
`!availableMemorySegments.isEmpty()` is not very necessary? Because it is duplicated with the condition in above `while (availableMemorySegments.isEmpty())`